### PR TITLE
Update to 6.0.0-preview1

### PR DESCRIPTION
**Pull request by aaalaniz**
### 6.0.0-preview1

* The Video Android SDK uses Chromium WebRTC 83.
* The Video Android SDK is built with NDK r20b.

#### Adopting WebRTC Classes

The update to video capturing and rendering includes an adoption of WebRTC classes defined in `libwebrtc.jar` included with the Video Android SDK. In previous major releases, the SDK would define public APIs with classes defined exclusively in the `com.twilio.video` package. The SDK will now include public APIs with classes defined in `libwebrtc.jar`.

Any public APIs that reference a WebRTC class are only compatible with the `libwebrtc.jar` provided from the Video Android SDK. The classes provided in this artifact are defined under the `tvi.webrtc` package to ensure side-by-side compatibility with other WebRTC based libraries.

Adopting WebRTC classses in public APIs provides developers already familiar with WebRTC more flexibility when capturing media and rendering media.

#### Video Capturing Update

This release includes an overhaul to the video capturer facilities. `LocalVideoTrack`s are now created with implementations of `tvi.webrtc.VideoCapturer` and `VideoFormat`s. `VideoCapturer` remains a part of the SDK as an extension of `tvi.webrtc.VideoCapturer`, but `VideoConstraints` have been completely removed in favor of `VideoFormat`s.

##### What's New!

* Leverage capturers provided by `libwebrtc.jar`
* Implement custom capturers that capture to textures
* Added support for NV12 format
* The video capturer API now enables cropping and scaling in response to network conditions

##### Updated Programming Model

When a caller create a `LocalVideoTrack`, the following methods of a `tvi.webrtc.VideoCapturer` are called in the order listed.

1. `isScreencast()` - The return value is used to create a `tvi.webrtc.VideoSource`
2. `initialize(surfaceTextureHelper, context, capturerObserver)` - This method is called so the capturer can setup to capture frames. The capturer should retain a reference to the capturer observer.
3. `startCapture(width, height, framerate)` - The capturer should start capturing in a format as close as possible to `width x height` at `framerate`.

When a caller releases a local video track, the following methods are then called.

1. `stopCapture()` - The capturer should stop capturing frames. The SDK expects this method to block until frames have stopped being captured.
2. `dispose()` - The capturer should perform any final clean up.

##### API Updates

* `VideoPixelFormat` has been removed in favor of implementing `tvi.webrtc.VideoFrame.Buffer`. The SDK includes support for the following buffer types: `tvi.webrtc.VideoFrame.TextureBuffer`, `tvi.webrtc.NV21Buffer`, `tvi.webrtc.NV12Buffer`, and `Rgba8888Buffer`.
* `VideoCapturer` now extends `tvi.webrtc.VideoCapturer`.
  * `getSupportedFormats()` has been removed in favor of `getCaptureFormat()`. This update allows a `VideoCapturer` to provide a default capture format. For example, the `ScreenCapturer` returns a capture format based on the device screen dimensions.
* `VideoConstraints` has been removed. Developers can now specify a capture format when creating a local video track and the `tvi.webrtc.VideoCapturer` is responsible for capturing at a format as close as possible to the specified format. The `VideoFormat` provided to `LocalVideoTrack.create(...)` takes highest priority over the format returned by `VideoCapturer#getCaptureFormat()`. The default format for video tracks remains 640x480 at 30 FPS.
* Removed `VideoFrame` in favor of `tvi.webrtc.VideoFrame`.
* Removed `AspectRatio`.

#### Video Rendering Update

This release includes updated video rendering facilities. `RemoteVideoTrack`s and `LocalVideoTrack`s are now rendered with a `tvi.webrtc.VideoSink` instead of a `VideoRenderer`. `VideoRenderer` and `VideoRenderer.Listener` have been removed in favor of `tvi.webrtc.VideoSink` and `tvi.webrtc.RendererCommon.RendererEvents` respectively.

##### API Updates

* Removed `VideoRenderer` in favor of `tvi.webrtc.VideoSink`
* Removed `VideoRenderer.Listener` in favor of `tvi.webrtc.RendererCommon.RendererEvents`
* Removed `VideoTrack#addRenderer` in favor of `VideoTrack#addSink`
* Removed `VideoTrack#removeRenderer` in favor of `VideoTrack#removeSink`
* Updated `VideoView#setListener` to take a `tvi.webrtc.RendererCommon.RendererEvents` instead of a `VideoRenderer.Listener`
* Updated `VideoTextureView#setListener` to take a `tvi.webrtc.RendererCommon.RendererEvents` instead of a `VideoRenderer.Listener`
* Removed `CameraCapturer#takePicture`. The functionality this API provides can be recreated using a custom `VideoSink`. Reference the [Advanced Camera Capturer Example](https://github.com/twilio/video-quickstart-android/tree/feature/6.0.0-preview/exampleAdvancedCameraCapturer) for more details.

##### SDK Size Update

The table below highlights the updated App Size Impact.

| ABI             | App Size Impact 5.x    | App Size Impact 6.0.0-preview1    |
| --------------- | ---------------------- |--------------------------|
| universal       | 21.9MB                 | 19.5MB                   |
| armeabi-v7a     | 4.8MB                  | 4.1MB                    |
| arm64-v8a       | 5.7MB                  | 5.1MB                    |
| x86             | 6MB                    | 5.5MB                    |
| x86_64          | 6.1MB                  | 5.4MB                    |

### CODE REVIEW

File: quickstartKotlin/src/main/java/com/twilio/video/quickstart/kotlin/SettingsActivity.kt [hunk 1]
```diff
@@ -100,8 +99,7 @@ class SettingsActivity : AppCompatActivity() {
                 VIDEO_CODEC_NAMES.toMutableList()
 
             // Remove H264 if not supported
-            if (!MediaCodecVideoDecoder.isH264HwSupported() ||
-                    !MediaCodecVideoEncoder.isH264HwSupported()) {
+            if (isH264Supported()) {
```
Comment (aaalaniz) [inline quickstartKotlin/src/main/java/com/twilio/video/quickstart/kotlin/SettingsActivity.kt:102]:
```suggestion
            if (!isH264Supported()) {
```

File: quickstart/src/main/java/com/twilio/video/quickstart/activity/SettingsActivity.java [hunk 2]
```diff
@@ -137,8 +135,7 @@ private void setupCodecListPreference(Class codecClass,
             }
 
             // Remove H264 if not supported
-            if (!MediaCodecVideoDecoder.isH264HwSupported() ||
-                    !MediaCodecVideoEncoder.isH264HwSupported()) {
+            if (CodecUtils.isH264Supported()) {
```
Comment (aaalaniz) [inline quickstart/src/main/java/com/twilio/video/quickstart/activity/SettingsActivity.java:138]:
```suggestion
            if (!CodecUtils.isH264Supported()) {
```

File: common/src/main/java/com/twilio/video/examples/common/VideoFrameKtx.kt [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+/**
+ * Converts a [tvi.webrtc.VideoFrame] to a Bitmap. This method must be called from a thread with a
+ * valid EGL context when the frame buffer is a [VideoFrame.TextureBuffer].
+ */
+fun VideoFrame.toBitmap() : Bitmap? {
+    return captureBitmap(this)
+}
+
+private fun captureBitmap(frame: VideoFrame) : Bitmap? {
```
Comment (Alton09) [inline common/src/main/java/com/twilio/video/examples/common/VideoFrameKtx.kt:26]: Nit pick but you can use a [single-expression function](https://kotlinlang.org/docs/reference/functions.html#single-expression-functions) here.

File: common/src/main/java/com/twilio/video/examples/common/VideoFrameKtx.kt [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    val rect =
+        Rect(0, 0, yuvImage.width, yuvImage.height)
+
+    // Compress YuvImage to jpeg
+    yuvImage.compressToJpeg(rect, 100, stream)
+
+    // Convert jpeg to Bitmap
+    val imageBytes = stream.toByteArray()
+    var bitmap: Bitmap?
+    bitmap = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
```
Comment (Alton09) [inline common/src/main/java/com/twilio/video/examples/common/VideoFrameKtx.kt:50]: We should be able to combine the declaration with the assignment into one statement and remove the type and rely on type inference here

File: exampleCustomVideoCapturer/src/main/java/com/twilio/video/examples/customcapturer/ViewCapturer.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            // Create video frame
+            val captureTimeNs =
+                TimeUnit.MILLISECONDS.toNanos(SystemClock.elapsedRealtime())
+            val videoBuffer: VideoFrame.Buffer =
+                Rgba8888Buffer(buffer, view.width, view.height)
+            val videoFrame = VideoFrame(videoBuffer, 0, captureTimeNs)
+
+            // Notify the observer
+            if (started.get()) {
+                capturerObserver!!.onFrameCaptured(videoFrame)
```
Comment (Alton09) [inline exampleCustomVideoCapturer/src/main/java/com/twilio/video/examples/customcapturer/ViewCapturer.kt:69]: Maybe use a null safe operator and throw a more detail exception if the reference is null?

**Review by Alton09 — CHANGES_REQUESTED** [review]

File: common/build.gradle [hunk 6]
```diff
@@ -0,0 +1,40 @@
+plugins {
+    id 'com.android.library'
+    id 'kotlin-android'
+}
+
+android {
+    compileSdkVersion versions.compileSdk
+
+    defaultConfig {
+        minSdkVersion 16
```
Comment (Alton09) [inline common/build.gradle:10]: Oh sorry, I forgot to mention the remaining hard coded values

File: common/build.gradle [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        targetSdkVersion 30
```
Comment (Alton09) [inline common/build.gradle:11]: Same as above

**Review by Alton09 — CHANGES_REQUESTED** [review]

**Review by Alton09 — APPROVED** [review]

**Pull request merged**